pkg/kubelet/kuberuntime/kuberuntime_logs.go

```diff
@@ -0,0 +1,332 @@
+				return fmt.Errorf("failed to reset seek: %v", err)
+			}
+			if w == nil {
+				// Intialize the watcher if it has not been initialized yet.
+				if w, err = fsnotify.NewWatcher(); err != nil {
+					return fmt.Errorf("failed to create fsnotify watcher: %v", err)
+				}
+				defer w.Close()
+				if err := w.Add(f.Name()); err != nil {
+					return fmt.Errorf("failed to watch file %q: %v", f.Name(), err)
+				}
+			}
+			// Wait until the next log change.
+			if err := waitLogs(w); err != nil {
+				return fmt.Errorf("failed to wait logs: %v", err)
+			}
+			continue
+		}
+		if parse == nil {
+			// Intialize the log parsing function.
+			parse, err = getParseFunc(l)
+			if err != nil {
+				return fmt.Errorf("failed to get parse function: %v", err)
+			}
+		}
+		// Parse the log line.
+		msg.reset()
+		if err := parse(l, msg); err != nil {
+			glog.Errorf("Failed with err %v when parsing log: %q", err, l)
+			continue
+		}
+		// Write the log line into the stream.
+		if err := writeLogs(msg, stdout, stderr, opts); err != nil {
+			glog.Errorf("Failed with err %v when writing log: %+v", err, msg)
+		}
+	}
+}
+
+// parseFunc is a function parsing one log line to the internal log type.
+// Notice that the caller must make sure logMessage is not nil.
+type parseFunc func([]byte, *logMessage) error
+
+var parseFuncs []parseFunc = []parseFunc{
+	parseCRILog,        // CRI log format parse function
+	parseDockerJSONLog, // Docker JSON log format parse function
+}
+
+// parseCRILog parses logs in CRI log format. CRI Log format example:
+//   2016-10-06T00:17:09.669794202Z stdout log content 1
+//   2016-10-06T00:17:09.669794203Z stderr log content 2
+func parseCRILog(log []byte, msg *logMessage) error {
+	var err error
+	// Parse timestamp
+	idx := bytes.Index(log, delimiter)
+	if idx < 0 {
+		return fmt.Errorf("timestamp is not found")
+	}
+	msg.timestamp, err = time.Parse(timeFormat, string(log[:idx]))
+	if err != nil {
+		return fmt.Errorf("unexpected timestamp format %q: %v", timeFormat, err)
+	}
+
+	// Parse stream type
+	log = log[idx+1:]
+	idx = bytes.Index(log, delimiter)
+	if idx < 0 {
+		return fmt.Errorf("stream type is not found")
+	}
+	msg.stream = streamType(log[:idx])
+	if msg.stream != stdoutType && msg.stream != stderrType {
+		return fmt.Errorf("unexpected stream type %q", msg.stream)
+	}
+
+	// Get log content
+	msg.log = log[idx+1:]
+
+	return nil
+}
+
+// dockerJSONLog is the JSON log buffer used in parseDockerJSONLog.
+var dockerJSONLog = &jsonlog.JSONLog{}
+
+// parseDockerJSONLog parses logs in Docker JSON log format. Docker JSON log format
+// example:
+//   {"log":"content 1","stream":"stdout","time":"2016-10-20T18:39:20.57606443Z"}
+//   {"log":"content 2","stream":"stderr","time":"2016-10-20T18:39:20.57606444Z"}
+func parseDockerJSONLog(log []byte, msg *logMessage) error {
+	dockerJSONLog.Reset()
+	l := dockerJSONLog
+	if err := json.Unmarshal(log, l); err != nil {
```

Comment:
This is hellishly expensive (I think).  Can you leave a TODO / open an issue to evaluate the performance of this later on?